### ISSUE TITLE
An out-of-the-box generator for pest grammar.

### ISSUE THREAD

**Issue by TheVeryDarkness** [pest-parser/ast#29]
I've written some codes for generating Concrete Syntax Tree from Pest grammar.

Maybe [that repository](https://github.com/theverydarkness/pest-typed) can be merged into some repository under [pest-parser](https://github.com/pest-parser) in the future if appreciated.

Examples can be found on [docs.rs](https://docs.rs/pest_typed_derive/latest/pest_typed_derive/) or [examples directory](https://github.com/TheVeryDarkness/pest-typed/tree/main/derive/examples).

An [issue](https://github.com/pest-parser/pest/issues/882) in pest suggests me to ask for feedback here :)

**Comment by TheVeryDarkness** [comment]
@tomtau I'll have some time one or two weeks later. I still have some exams and assignments now. And I guess it won't take too long to implement them.
I've also read those discussions before. 3 points you listed above are definitely part of what I like to see. And I think we're making pest a modern parser.

**Comment by marcfir** [comment]
Can I jump in here and ask some general questions about ASTs in Rust and pest?

From my point of view, there are two important questions about generating an AST from the grammar.
1. How do we want to do the abstraction?
2. What kind of data structure do we want to use?

In general, the question is who should generate the AST. Because an AST is always an abstracted view of the parsing result. So, depending on what you want to do, your AST may look different and may be implemented differently.
But we can look at a generic use of parser generators. These are language workbenches such as [xtext](https://eclipse.dev/Xtext/index.html), [langium](https://langium.org/), or [monticore](https://monticore.github.io/monticore/).

### AST abstraction
If you look at the language workbenches mentioned above, they typically do the following when translated to pest:

 - A **sequence** like `entry = { SOI ~ expr ~ number ~ EOI }` is mapped to single struct fields `struct Entry {expr: Expr, Number: Number}`. More precisely, the fields would be wrapped in a `Box` to support expressions.
 - An **ordered choice** like `operation = { add | subtract | multiply | divide | power }` is mapped to multiple optional fields `struct operation {add: Option<Add>, Subtract: Option<Subtract>, Multiply: Option<Multiply>, divide: Option<Divide>, power: Option<Power> }`. 
 - A **Repetition** like `expr = { ( term)* }` is mapped to a vector `struct Expr{term:Vec<Term>}`.

This would be somewhat simpler than the current procedure in [TheVeryDarkness/pest-typed](https://github.com/TheVeryDarkness/pest-typed)

### Data Structure
One problem in Rust is references to AST nodes. In other dynamic languages it is easy to have references inside a tree structure and to three nodes. But in Rust it is more complicated, especially if you have mutable references.
So the first question might be, why do we need references? There are e.g. use cases:
- **Validation** We typically want to validate the semantics of our language. This is usually done with a visitor pattern. A nice API will give us access to the parent node.
- **Symboling** We typically have symbols/identifiers in our language that are handled in a symbol table. We want to reference the AST node from a particular symbol.

So we could construct our AST based on `Rc<RefCell<...>>` or we could use an arena allocator based approach with a flat vector where the references are indicated by the index in the array. The arena AST is very simple in referencing and lifetime. Also the mutability for AST nodes is easier.

We experimented with all the approaches and ended up with an arena AST.

@TheVeryDarkness What were your thoughts for your approache?

**Comment by TheVeryDarkness** [comment]
There's a PR <https://github.com/pest-parser/pest/pull/815> about hooks.

**Comment by marcfir** [comment]
@TheVeryDarkness The problem with a rule like this `row = { (item ~ comma)* ~ item }` can be solved with the tagging feature invented with pest v2.6.0. So you can write this `row = { (item ~ comma)* ~ #last_item = item }` which will be translated to something like this `struct Row { item: Vec<Item>, comma: Vec<Comma>, last_item: Item}`.

I also hit the problem, that macros cannot share state. There is a related issue rust-lang/rust#44034.

I am currently working on a language workbench in Rust based on pest. It's not feature complete and very unstable yet. So we haven't released it yet. Our first start was an automatic AST generation. So I think it makes sense to at least show the AST generation. I released the code at [pestast](https://github.com/marcfir/pestast). If I have more time I can also commit the arena tree based AST

@tomtau I didn't have the discussions in mind. So maybe we can continue there at least for pest3.
And I agree with your event-based approach. It makes it very modular and there could be multiple libraries providing different types of ASTs.

**Comment by TheVeryDarkness** [comment]
@marcfir Yes, it's practicable if rule components are well tagged (though tags are available only under feature *grammar-extras*). I think we can provide typed AST in both styles.

By the way, `pest-typed` also supports accessing using rule names. `row.last_item()` is also available, and there's a attribute that controls whether one can access rules or tags inside a tag from outside directly.

**Comment by TheVeryDarkness** [comment]
> @TheVeryDarkness FYI regarding trivia, currently the intended semantics / usage is this:
> 
> 1. a user can define / override these special rules:
> 
> ```
> ~ = (whitespace | comment)*
> ^ = (whitespace | comment)*
> ```
> 
> 2. there are then 3 types of sequencing operators:
> 
> A) `-`: no implicit whitespace/trivia (just like atomic rules in pest 2, but for a single sequence) B) `~`: optional implicit whitespace C) `^`: mandatory implicit whitespace
> 
> 3. based on those implicit whitespace sequence operators, there are also corresponding infix, e.g.:
> 
> A) `rule*`: rule repeated without implicit whitespace/trivia B) `rule~*`: rule repeated with optional implicit whitespace/trivia C) `rule^*`: rule repeated with mandatory implicit whitespace/trivia

Sorry, I have some questions about the definition of trivia:

```pest3
~ = (whitespace | comment)*
^ = (whitespace | comment)*
```

Are there any differences between them? And should both of them be defined in the same grammar?

I though there is only one form of trivia, and there should be only one way to define the trivia.

**Comment by TheVeryDarkness** [comment]
And one issue is whether we should keep skipped trivia inside our generated data structure? It may be useful in some cases, such as implementing a formatter, but can also be solved by defining the rule explicitly. So, I may drop skipped trivia in my PR, unlike what I've done in `pest-typed`.

**Comment by TheVeryDarkness** [comment]
> > Are there any differences between them? And should (or could) both of them be defined in the same grammar?
> 
> Yes, it's mentioned below; ~ means that there's an optional trivia (0 or more) between items, ^ means there's a mandatory trivia (1 or more). It's currently unspecified whether we'd always want both to be defined. I guess we could mimic pest2 where WHITESPACE and COMMENT don't need to be defined?

Does it mean we should write codes like below?

```pest3
~ = { (whitespace | comment)* }
^ = { (whitespace | comment)+ }
```

**Comment by TheVeryDarkness** [comment]
I have a fork [here](https://github.com/TheVeryDarkness/pest3) and I'm working on it.
I'm sorry for my slow progress. I didn't pay enough time on it.
